File: app/src/main/java/net/czlee/debatekeeper/DebateFormatDownloadManager.java
/*
 * Copyright (C) 2021 Chuan-Zheng Lee
 *
 * This file is part of the Debatekeeper app, which is licensed under the GNU
 * General Public Licence version 3 (GPLv3).  You can redistribute and/or modify
 * it under the terms of the GPLv3, and you must not use this file except in
 * compliance with the GPLv3.
 *
 * This app is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU General Public Licence for more details.
 *
 * You should have received a copy of the GNU General Public Licence along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.czlee.debatekeeper;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;
import android.util.JsonReader;
import android.util.Log;
import android.util.MalformedJsonException;

import androidx.annotation.NonNull;
import androidx.core.os.HandlerCompat;
import androidx.preference.PreferenceManager;

import net.czlee.debatekeeper.debateformat.DebateFormatFieldExtractor;
import net.czlee.debatekeeper.debateformat.LanguageChooser;

import org.xml.sax.SAXException;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Manager for downloading information and format XML files from the Debatekeeper formats server.
 * This class:
 * <ul>
 *     <li>contacts the server to download a list of all format XML files available</li>
 *     <li>provides a class representing a downloadable format for {@link DownloadableFormatRecyclerAdapter}</li>
 *     <li>downloads requested format XML files and saves them to the user's device</li>
 * </ul>
 *
 * @author Chuan-Zheng Lee
 * @since 2021-09-28
 */
public class DebateFormatDownloadManager {

    private final String TAG = "DebateFormatDownload";

    private final Context mContext;
    private final ArrayList<DownloadableFormatEntry> mEntries;
    private final DownloadFormatsFragment.DownloadBinder mBinder;
    private ExecutorService mExecutorService;
    private Handler mMainThreadHandler;
    private final FormatXmlFilesManager mFilesManager;

    //******************************************************************************************
    // Public class
    //******************************************************************************************

    public static class DownloadableFormatEntry implements Comparable<DownloadableFormatEntry> {

        public enum DownloadState {
            NOT_DOWNLOADED,
            UPDATE_AVAILABLE,
            DOWNLOAD_IN_PROGRESS,
            DOWNLOADED
        }

        private static final String TAG = "DownloadFormatEntry";

        public int version = 0;
        @NonNull
        public String filename = "";
        @NonNull
        public String url = "";
        @NonNull
        public String name = "";
        @NonNull
        public String[] regions = new String[0];
        @NonNull
        public String[] usedAts = new String[0];
        @NonNull
        public String[] levels = new String[0];
        @NonNull
        public String description = "";
        @NonNull
        public DownloadState state = DownloadState.NOT_DOWNLOADED;
        public boolean expanded = true;

        /**
         * Given a format XML files manager and version extractor, checks the version number in the
         * existing file in the files manager (if any) and updates the <code>state</code> field
         * accordingly.
         *
         * @param filesManager     a {@link FormatXmlFilesManager}
         * @param versionExtractor a {@link DebateFormatFieldExtractor} that should be initialised
         *                         with <code>fieldName</code> as the version field.
         */
        void checkForExistingFile(FormatXmlFilesManager filesManager, DebateFormatFieldExtractor versionExtractor) {
            InputStream in;
            String versionStr;

            if (!filesManager.exists(this.filename)) {
                this.state = DownloadState.NOT_DOWNLOADED;
                return;
            }

            try {
                in = filesManager.open(this.filename);
                versionStr = versionExtractor.getFieldValue(in);
            } catch (IOException | SAXException e) {
                Log.e(TAG, "Couldn't get version from " + this.filename);
                this.state = DownloadState.NOT_DOWNLOADED;
                return;
            }

            int existingVersion;
            try {
                existingVersion = Integer.parseInt(versionStr);
            } catch (NumberFormatException e) {
                Log.e(TAG, "Invalid version in " + this.filename + ": " + versionStr);
                this.state = DownloadState.UPDATE_AVAILABLE;
                return;
            }

            if (this.version > existingVersion)
                this.state = DownloadState.UPDATE_AVAILABLE;
            else
                this.state = DownloadState.DOWNLOADED;
        }

        @Override
        public int compareTo(DownloadableFormatEntry o) {
            // when sorting, sort by name
            return this.name.compareTo(o.name);
        }

    }

    //******************************************************************************************
    // Private class
    //******************************************************************************************

    /**
     * Helper class that builds a {@link DownloadableFormatEntry} using a {@link JsonReader}. This
     * class mostly just helps to group these functions together, but it does maintain state in a
     * {@link LanguageChooser}.
     */
    private static class DownloadableFormatListBuilder {

        private static class FormatInfo {
            @NonNull
            public String name = "";
            @NonNull
            public String[] regions = new String[0];
            @NonNull
            public String[] usedAts = new String[0];
            @NonNull
            public String[] levels = new String[0];
            @NonNull
            public String description = "";
        }

        private final LanguageChooser mLangChooser;

        DownloadableFormatListBuilder() {
            mLangChooser = new LanguageChooser();
        }

        /**
         * Entry function. Builds a list of {@link DownloadableFormatEntry} objects from the given
         * {@link JsonReader}.
         *
         * See <a href="https://github.com/czlee/debatekeeper-formats/blob/main/v1/formats.json">
         * formats.json in the official formats repository</a> for an example of a
         * JSON file this expects to parse.
         *
         * @param reader a {@link JsonReader} pointing to the (start of a) JSON file
         * @return list of {@link DownloadableFormatEntry} objects
         * @throws IOException if there is an issue parsing the JSON
         */
        List<DownloadableFormatEntry> buildListFromJson(JsonReader reader) throws IOException {
            ArrayList<DownloadableFormatEntry> entries = new ArrayList<>();

            reader.beginObject();
            while (reader.hasNext()) {
                String key = reader.nextName();
                if (key.equals("formats"))
                    addFormatsFromJson(reader, entries);
                else
                    reader.skipValue();
            }
            reader.endObject();

            return entries;
        }

        /**
         * Adds entries to the given list from the given {@link JsonReader}. Expects to find an
         * array containing entry objects in the JSON file.
         */
        void addFormatsFromJson(JsonReader reader, List<DownloadableFormatEntry> entries) throws IOException {
            reader.beginArray();
            while (reader.hasNext()) {
                DownloadableFormatEntry entry = buildEntryFromJson(reader);
                // Log.d(TAG, "added: " + entry.styleName + " (" + entry.filename + ")");
                entries.add(entry);
            }
            reader.endArray();
        }


        /**
         * Builds a {@link DownloadableFormatEntry} from a {@link JsonReader}. Expects to find an
         * object representing an entry in the JSON file.
         */
        DownloadableFormatEntry buildEntryFromJson(JsonReader reader) throws IOException, IllegalStateException {
            DownloadableFormatEntry entry = new DownloadableFormatEntry();
            reader.beginObject();
            while (reader.hasNext()) {
                String key = reader.nextName();
                switch (key) {
                    case "filename":
                        entry.filename = reader.nextString();
                        break;
                    case "info":
                        Map<String, FormatInfo> infoObjects = readInfoObjects(reader);
                        FormatInfo info = chooseInfo(infoObjects);
                        if (info != null) {
                            entry.name = info.name;
                            entry.regions = info.regions;
                            entry.levels = info.levels;
                            entry.usedAts = info.usedAts;
                            entry.description = info.description;
                        }
                        break;
                    case "version":
                        entry.version = reader.nextInt();
                        break;
                    case "url":
                        entry.url = reader.nextString();
                        break;
                    default:
                        reader.skipValue();
                }
            }
            reader.endObject();
            return entry;
        }

        /**
         * Reads the "info" object for all languages.
         *
         * @param reader a {@link JsonReader}
         * @return a {@link Map} mapping (possibly empty) language codes to {@link FormatInfo}
         * objects for that language.
         */
        Map<String, FormatInfo> readInfoObjects(JsonReader reader) throws IOException {
            HashMap<String, FormatInfo> infoObjects = new HashMap<>();
            reader.beginObject();
            while (reader.hasNext()) {
                String lang = reader.nextName();
                FormatInfo info = new FormatInfo();
                reader.beginObject();
                while (reader.hasNext()) {
                    String key = reader.nextName();
                    switch (key) {
                        case "name":
                            info.name = reader.nextString();
                            break;
                        case "regions":
                            info.regions = readStringList(reader);
                            break;
                        case "levels":
                            info.levels = readStringList(reader);
                            break;
                        case "used-ats":
                            info.usedAts = readStringList(reader);
                            break;
                        case "description":
                            info.description = reader.nextString();
                            break;
                        default:
                            reader.skipValue();
                    }
                }
                reader.endObject();
                infoObjects.put(lang, info);
            }
            reader.endObject();
            return infoObjects;
        }

        FormatInfo chooseInfo(Map<String, FormatInfo> infoObjects) {
            List<String> languages = new ArrayList<>(infoObjects.keySet());
            String chosen = mLangChooser.choose(languages);
            if (chosen == null) return null;
            else return infoObjects.get(chosen);
        }

        String[] readStringList(JsonReader reader) throws IOException, IllegalStateException {
            ArrayList<String> strings = new ArrayList<>();
            reader.beginArray();
            while (reader.hasNext()) {
                strings.add(reader.nextString());
            }
            reader.endArray();
            return strings.toArray(new String[0]);
        }

    }

    //******************************************************************************************
    // Public constructor
    //******************************************************************************************

    DebateFormatDownloadManager(Context context, DownloadFormatsFragment.DownloadBinder binder) {
        mContext = context;
        mBinder = binder;
        mEntries = new ArrayList<>();
        mFilesManager = new FormatXmlFilesManager(context);
    }

    //******************************************************************************************
    // Public methods
    //******************************************************************************************

    public ArrayList<DownloadableFormatEntry> getEntries() {
        return mEntries;
    }

    /**
     * Starts downloading the list of downloadable entries from the server, and notifies the binder
     * on the main thread when done. This should be safe to call from any thread.
     */
    public void startDownloadList() {
        initialiseThreads();
        mExecutorService.execute(() -> {
            try {
                List<DownloadableFormatEntry> newEntries = synchronousDownloadList();
                mMainThreadHandler.post(() -> replaceEntriesAndNotify(newEntries));
            } catch (MalformedJsonException e) {
                e.printStackTrace();
                String originalMessage = e.getMessage();
                String message = originalMessage;

                // Since we're using a strict JsonReader, the error is normally "Use JsonReader.setLenient(true)...".
                // If this is the case, try to extract the line and column number.
                if (originalMessage != null && originalMessage.startsWith("Use JsonReader.setLenient")) {
                    Matcher matcher = Pattern.compile("line \\d+ column \\d+").matcher(originalMessage);
                    if (matcher.find())
                        message = "Malformed JSON at " + matcher.group();
                }

                final String finalMessage = message;
                mMainThreadHandler.post(() -> mBinder.notifyJsonParseError(finalMessage));
            } catch (IOException e) {
                e.printStackTrace();
                String message = (e instanceof FileNotFoundException)
                        ? mContext.getString(R.string.formatDownloader_notFoundError, e.getLocalizedMessage())
                        : e.getLocalizedMessage();
                mMainThreadHandler.post(() -> mBinder.notifyListDownloadError(message));
            } catch (IllegalStateException | NumberFormatException e) {
                e.printStackTrace();
                mMainThreadHandler.post(() -> mBinder.notifyJsonParseError(e.getLocalizedMessage()));
            }
        });
    }

    /**
     * Starts downloading the format file represented by the given {@link DownloadableFormatEntry},
     * calling <code>holder.updateDownloadProgress()</code> on the main thread as it starts and
     * completes. This must be called from the main thread.
     *
     * @param entry  a {@link DownloadableFormatEntry}
     * @param holder a {@link DownloadableFormatRecyclerAdapter.ViewHolder}
     */
    public void startDownloadFile(DownloadableFormatEntry entry, DownloadableFormatRecyclerAdapter.ViewHolder holder) {
        initialiseThreads();
        final DownloadableFormatEntry.DownloadState originalState = entry.state;
        entry.state = DownloadableFormatEntry.DownloadState.DOWNLOAD_IN_PROGRESS;
        holder.updateDownloadProgress();
        mExecutorService.execute(() -> {
            try {
                synchronousDownloadFile(entry);
                mMainThreadHandler.post(() -> {
                    entry.state = DownloadableFormatEntry.DownloadState.DOWNLOADED;
                    holder.updateDownloadProgress();
                });
            } catch (IOException e) {
                e.printStackTrace();
                String message = (e instanceof FileNotFoundException)
                        ? mContext.getString(R.string.formatDownloader_notFoundError, e.getLocalizedMessage())
                        : e.getLocalizedMessage();
                mMainThreadHandler.post(() -> {
                    entry.state = originalState;
                    holder.updateDownloadProgress();
                    mBinder.showSnackbarError(entry.filename, message);
                });
            }
        });
    }

    //******************************************************************************************
    // Private methods
    //******************************************************************************************

    /**
     * Initialises thread management if it hasn't already been initialised. Must be called before
     * any background thread is done. Does nothing if initialisation has already happened.
     */
    private void initialiseThreads() {
        if (mExecutorService == null)
            mExecutorService = Executors.newSingleThreadExecutor();
        if (mMainThreadHandler == null)
            mMainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper());
    }

    /**
     * Replaces <code>mEntries</code> with <code>newEntries</code>, and notifies the binder that the
     * data has changed. Must be run on the main thread.
     *
     * @param newEntries the new entries
     */
    private void replaceEntriesAndNotify(List<DownloadableFormatEntry> newEntries) {
        int originalSize = mEntries.size();
        mEntries.clear();
        mEntries.addAll(newEntries);
        mBinder.notifyAdapterItemsReplaced(originalSize, newEntries.size());
    }

    /**
     * Downloads the file represented by a {@link DownloadableFormatEntry}. This accesses the
     * network, so it must be run on a background thread. It also checks that the URL host matches
     * that of where the format came from.
     *
     * @param entry a {@link DownloadableFormatEntry}
     */
    private void synchronousDownloadFile(DownloadableFormatEntry entry) throws IOException {
        Log.i(TAG, "Downloading file from server: " + entry.filename);

        URL url = new URL(entry.url);
        Log.d(TAG, "url: " + url.toString());
        if (!verifyHostMatch(url)) {
            String message = mContext.getString(R.string.formatDownloader_wrongHostError, url.toString());
            throw new MalformedURLException(message);
        }

        URLConnection connection = url.openConnection();
        InputStream in = connection.getInputStream();

        mFilesManager.copy(in, entry.filename);
    }

    /**
     * Downloads the list of formats from the server. This accesses the network, so it must be run
     * on a background thread. The list it returns is unmodifiable, to try to protect against
     * accidental threading errors.
     */
    private List<DownloadableFormatEntry> synchronousDownloadList()
            throws IOException, IllegalStateException, NumberFormatException {
        Log.i(TAG, "Downloading list from server");

        URL url = getListUrl();
        URLConnection connection = url.openConnection();
        InputStream in = connection.getInputStream();
        InputStreamReader isr = new InputStreamReader(in);

        DownloadableFormatListBuilder listBuilder = new DownloadableFormatListBuilder();
        List<DownloadableFormatEntry> entries;

        try (JsonReader reader = new JsonReader(isr)) {
            entries = listBuilder.buildListFromJson(reader);
        } catch (IOException e) {
            return Collections.emptyList();
        }

        // Check for available updates
        DebateFormatFieldExtractor versionExtractor = new DebateFormatFieldExtractor(mContext, R.string.xml2elemName_version);
        for (DownloadableFormatEntry entry : entries)
            entry.checkForExistingFile(mFilesManager, versionExtractor);

        Collections.sort(entries);
        return Collections.unmodifiableList(entries);
    }

    private boolean verifyHostMatch(URL url) throws MalformedURLException {
        URL listUrl = getListUrl();
        return url.getHost().equals(listUrl.getHost());
    }

    private URL getListUrl() throws MalformedURLException {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(mContext);
        String userUrl = prefs.getString(DownloadFormatsFragment.PREFERENCE_DOWNLOAD_LIST_URL, null);
        if (userUrl == null) {
            String defaultUrl = mContext.getString(R.string.formatDownloader_list_defaultUrl);
            Log.i(TAG, "using default URL: " + defaultUrl);
            return new URL(defaultUrl);
        } else {
            Log.i(TAG, "using user-provided URL: " + userUrl);
            return new URL(userUrl);
        }
    }

}

File: app/src/main/java/net/czlee/debatekeeper/DebateFormatDownloadManager.kt
/*
 * Copyright (C) 2021 Chuan-Zheng Lee
 *
 * This file is part of the Debatekeeper app, which is licensed under the GNU
 * General Public Licence version 3 (GPLv3).  You can redistribute and/or modify
 * it under the terms of the GPLv3, and you must not use this file except in
 * compliance with the GPLv3.
 *
 * This app is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU General Public Licence for more details.
 *
 * You should have received a copy of the GNU General Public Licence along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.czlee.debatekeeper

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.JsonReader
import android.util.Log
import android.util.MalformedJsonException
import androidx.core.os.HandlerCompat
import androidx.preference.PreferenceManager
import net.czlee.debatekeeper.debateformat.DebateFormatFieldExtractor
import net.czlee.debatekeeper.debateformat.LanguageChooser
import org.xml.sax.SAXException
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.MalformedURLException
import java.net.URL
import java.util.Collections
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.regex.Pattern

/**
 * Manager for downloading information and format XML files from the Debatekeeper formats server.
 * This class:
 *
 *  - contacts the server to download a list of all format XML files available
 *  - provides a class representing a downloadable format for [DownloadableFormatRecyclerAdapter]
 *  - downloads requested format XML files and saves them to the user's device
 *
 * @author Chuan-Zheng Lee
 * @since 2021-09-28
 */
class DebateFormatDownloadManager internal constructor(
        private val mContext: Context,
        private val mBinder: DownloadFormatsFragment.DownloadBinder) {

    private val TAG = "DebateFormatDownload"

    val entries = ArrayList<DownloadableFormatEntry>()
    private var mExecutorService: ExecutorService? = null
    private var mMainThreadHandler: Handler? = null
    private val mFilesManager = FormatXmlFilesManager(mContext)

    //******************************************************************************************
    // Public class
    //******************************************************************************************

    class DownloadableFormatEntry : Comparable<DownloadableFormatEntry> {

        enum class DownloadState {
            NOT_DOWNLOADED,
            UPDATE_AVAILABLE,
            DOWNLOAD_IN_PROGRESS,
            DOWNLOADED
        }

        @JvmField
        var version = 0
        @JvmField
        var filename = ""
        @JvmField
        var url = ""
        @JvmField
        var name = ""
        @JvmField
        var regions = arrayOf<String>()
        @JvmField
        var usedAts = arrayOf<String>()
        @JvmField
        var levels = arrayOf<String>()
        @JvmField
        var description = ""
        @JvmField
        var state = DownloadState.NOT_DOWNLOADED
        @JvmField
        var expanded = false

        /**
         * Given a format XML files manager and version extractor, checks the version number in the
         * existing file in the files manager (if any) and updates the [state] field
         * accordingly.
         *
         * @param filesManager     a [FormatXmlFilesManager]
         * @param versionExtractor a [DebateFormatFieldExtractor] that should be initialised
         *                         with `fieldName` as the version field.
         */
        internal fun checkForExistingFile(filesManager: FormatXmlFilesManager,
                                          versionExtractor: DebateFormatFieldExtractor) {
            if (!filesManager.exists(filename)) {
                state = DownloadState.NOT_DOWNLOADED
                return
            }

            val versionStr: String?
            try {
                val `in` = filesManager.open(filename)
                versionStr = versionExtractor.getFieldValue(`in`)
            } catch (e: IOException) {
                Log.e(TAG, "Couldn't get version from $filename")
                state = DownloadState.NOT_DOWNLOADED
                return
            } catch (e: SAXException) {
                Log.e(TAG, "Couldn't get version from $filename")
                state = DownloadState.NOT_DOWNLOADED
                return
            }

            if (versionStr == null) {
                Log.e(TAG, "No version found in $filename")
                state = DownloadState.UPDATE_AVAILABLE
                return
            }

            val existingVersion: Int
            try {
                existingVersion = versionStr.toInt()
            } catch (e: NumberFormatException) {
                Log.e(TAG, "Invalid version in $filename: $versionStr")
                state = DownloadState.UPDATE_AVAILABLE
                return
            }

            state = if (version > existingVersion)
                DownloadState.UPDATE_AVAILABLE
            else
                DownloadState.DOWNLOADED
        }

        override fun compareTo(other: DownloadableFormatEntry): Int {
            // when sorting, sort by name
            return name.compareTo(other.name)
        }

        companion object {
            private const val TAG = "DownloadFormatEntry"
        }
    }

    //******************************************************************************************
    // Private class
    //******************************************************************************************

    /**
     * Helper class that builds a [DownloadableFormatEntry] using a [JsonReader]. This
     * class mostly just helps to group these functions together, but it does maintain state in a
     * [LanguageChooser].
     */
    private class DownloadableFormatListBuilder {

        private class FormatInfo {
            var name = ""
            var regions = arrayOf<String>()
            var usedAts = arrayOf<String>()
            var levels = arrayOf<String>()
            var description = ""
        }

        private val mLangChooser = LanguageChooser()

        /**
         * Entry function. Builds a list of [DownloadableFormatEntry] objects from the given
         * [JsonReader].
         *
         * See [formats.json in the official formats repository](https://github.com/czlee/debatekeeper-formats/blob/main/v1/formats.json)
         * for an example of a JSON file this expects to parse.
         *
         * @param reader a [JsonReader] pointing to the (start of a) JSON file
         * @return list of [DownloadableFormatEntry] objects
         * @throws IOException if there is an issue parsing the JSON
         */
        @Throws(IOException::class)
        fun buildListFromJson(reader: JsonReader): MutableList<DownloadableFormatEntry> {
            val entries = ArrayList<DownloadableFormatEntry>()

            reader.beginObject()
            while (reader.hasNext()) {
                val key = reader.nextName()
                if (key == "formats")
                    addFormatsFromJson(reader, entries)
                else
                    reader.skipValue()
            }
            reader.endObject()

            return entries
        }

        /**
         * Adds entries to the given list from the given [JsonReader]. Expects to find an
         * array containing entry objects in the JSON file.
         */
        @Throws(IOException::class)
        fun addFormatsFromJson(reader: JsonReader, entries: MutableList<DownloadableFormatEntry>) {
            reader.beginArray()
            while (reader.hasNext()) {
                val entry = buildEntryFromJson(reader)
                // Log.d(TAG, "added: " + entry.styleName + " (" + entry.filename + ")");
                entries.add(entry)
            }
            reader.endArray()
        }

        /**
         * Builds a [DownloadableFormatEntry] from a [JsonReader]. Expects to find an
         * object representing an entry in the JSON file.
         */
        @Throws(IOException::class, IllegalStateException::class)
        fun buildEntryFromJson(reader: JsonReader): DownloadableFormatEntry {
            val entry = DownloadableFormatEntry()
            reader.beginObject()
            while (reader.hasNext()) {
                when (reader.nextName()) {
                    "filename" -> entry.filename = reader.nextString()
                    "info" -> {
                        val infoObjects = readInfoObjects(reader)
                        val info = chooseInfo(infoObjects)
                        if (info != null) {
                            entry.name = info.name
                            entry.regions = info.regions
                            entry.levels = info.levels
                            entry.usedAts = info.usedAts
                            entry.description = info.description
                        }
                    }
                    "version" -> entry.version = reader.nextInt()
                    "url" -> entry.url = reader.nextString()
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
            return entry
        }

        /**
         * Reads the "info" object for all languages.
         *
         * @param reader a [JsonReader]
         * @return a [Map] mapping (possibly empty) language codes to [FormatInfo]
         * objects for that language.
         */
        @Throws(IOException::class)
        fun readInfoObjects(reader: JsonReader): Map<String, FormatInfo> {
            val infoObjects = HashMap<String, FormatInfo>()
            reader.beginObject()
            while (reader.hasNext()) {
                val lang = reader.nextName()
                val info = FormatInfo()
                reader.beginObject()
                while (reader.hasNext()) {
                    when (reader.nextName()) {
                        "name" -> info.name = reader.nextString()
                        "regions" -> info.regions = readStringList(reader)
                        "levels" -> info.levels = readStringList(reader)
                        "used-ats" -> info.usedAts = readStringList(reader)
                        "description" -> info.description = reader.nextString()
                        else -> reader.skipValue()
                    }
                }
                reader.endObject()
                infoObjects[lang] = info
            }
            reader.endObject()
            return infoObjects
        }

        fun chooseInfo(infoObjects: Map<String, FormatInfo>): FormatInfo? {
            val languages = ArrayList(infoObjects.keys)
            val chosen = mLangChooser.choose(languages) ?: return null
            return infoObjects[chosen]
        }

        @Throws(IOException::class, IllegalStateException::class)
        fun readStringList(reader: JsonReader): Array<String> {
            val strings = ArrayList<String>()
            reader.beginArray()
            while (reader.hasNext()) {
                strings.add(reader.nextString())
            }
            reader.endArray()
            return strings.toTypedArray()
        }
    }

    //******************************************************************************************
    // Public methods
    //******************************************************************************************

    /**
     * Starts downloading the list of downloadable entries from the server, and notifies the binder
     * on the main thread when done. This should be safe to call from any thread.
     */
    fun startDownloadList() {
        initialiseThreads()
        mExecutorService!!.execute {
            try {
                val newEntries = synchronousDownloadList()
                mMainThreadHandler!!.post { replaceEntriesAndNotify(newEntries) }
            } catch (e: MalformedJsonException) {
                e.printStackTrace()
                val originalMessage = e.message
                var message = originalMessage

                // Since we're using a strict JsonReader, the error is normally "Use
                // JsonReader.setLenient(true)...". If this is the case, try to extract the line
                // and column number.
                if (originalMessage != null && originalMessage.startsWith("Use JsonReader.setLenient")) {
                    val matcher = Pattern.compile("line \\d+ column \\d+").matcher(originalMessage)
                    if (matcher.find())
                        message = "Malformed JSON at " + matcher.group()
                }

                val finalMessage = message
                mMainThreadHandler!!.post { mBinder.notifyJsonParseError(finalMessage) }
            } catch (e: IOException) {
                e.printStackTrace()
                val message = if (e is FileNotFoundException)
                    mContext.getString(R.string.formatDownloader_notFoundError, e.localizedMessage)
                else
                    e.localizedMessage
                mMainThreadHandler!!.post { mBinder.notifyListDownloadError(message) }
            } catch (e: IllegalStateException) {
                e.printStackTrace()
                mMainThreadHandler!!.post { mBinder.notifyJsonParseError(e.localizedMessage) }
            } catch (e: NumberFormatException) {
                e.printStackTrace()
                mMainThreadHandler!!.post { mBinder.notifyJsonParseError(e.localizedMessage) }
            }
        }
    }

    /**
     * Starts downloading the format file represented by the given [DownloadableFormatEntry],
     * calling `holder.updateDownloadProgress()` on the main thread as it starts and
     * completes. This must be called from the main thread.
     *
     * @param entry  a [DownloadableFormatEntry]
     * @param holder a [DownloadableFormatRecyclerAdapter.ViewHolder]
     */
    fun startDownloadFile(entry: DownloadableFormatEntry,
                          holder: DownloadableFormatRecyclerAdapter.ViewHolder) {
        initialiseThreads()
        val originalState = entry.state
        entry.state = DownloadableFormatEntry.DownloadState.DOWNLOAD_IN_PROGRESS
        holder.updateDownloadProgress()
        mExecutorService!!.execute {
            try {
                synchronousDownloadFile(entry)
                mMainThreadHandler!!.post {
                    entry.state = DownloadableFormatEntry.DownloadState.DOWNLOADED
                    holder.updateDownloadProgress()
                }
            } catch (e: IOException) {
                e.printStackTrace()
                val message = if (e is FileNotFoundException)
                    mContext.getString(R.string.formatDownloader_notFoundError, e.localizedMessage)
                else
                    e.localizedMessage
                mMainThreadHandler!!.post {
                    entry.state = originalState
                    holder.updateDownloadProgress()
                    mBinder.showSnackbarError(entry.filename, message)
                }
            }
        }
    }

    //******************************************************************************************
    // Private methods
    //******************************************************************************************

    /**
     * Initialises thread management if it hasn't already been initialised. Must be called before
     * any background thread is done. Does nothing if initialisation has already happened.
     */
    private fun initialiseThreads() {
        if (mExecutorService == null)
            mExecutorService = Executors.newSingleThreadExecutor()
        if (mMainThreadHandler == null)
            mMainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper())
    }

    /**
     * Replaces [entries] with `newEntries`, and notifies the binder that the
     * data has changed. Must be run on the main thread.
     *
     * @param newEntries the new entries
     */
    private fun replaceEntriesAndNotify(newEntries: List<DownloadableFormatEntry>) {
        val originalSize = entries.size
        entries.clear()
        entries.addAll(newEntries)
        mBinder.notifyAdapterItemsReplaced(originalSize, newEntries.size)
    }

    /**
     * Downloads the file represented by a [DownloadableFormatEntry]. This accesses the
     * network, so it must be run on a background thread. It also checks that the URL host matches
     * that of where the format came from.
     *
     * @param entry a [DownloadableFormatEntry]
     */
    @Throws(IOException::class)
    private fun synchronousDownloadFile(entry: DownloadableFormatEntry) {
        Log.i(TAG, "Downloading file from server: " + entry.filename)

        val url = URL(entry.url)
        Log.d(TAG, "url: $url")
        if (!verifyHostMatch(url)) {
            val message = mContext.getString(R.string.formatDownloader_wrongHostError,
                    url.toString())
            throw MalformedURLException(message)
        }

        val connection = url.openConnection()
        val `in` = connection.getInputStream()

        mFilesManager.copy(`in`, entry.filename)
    }

    /**
     * Downloads the list of formats from the server. This accesses the network, so it must be run
     * on a background thread. The list it returns is unmodifiable, to try to protect against
     * accidental threading errors.
     */
    @Throws(IOException::class, IllegalStateException::class, NumberFormatException::class)
    private fun synchronousDownloadList(): List<DownloadableFormatEntry> {
        Log.i(TAG, "Downloading list from server")

        val url = getListUrl()
        val connection = url.openConnection()
        val `in`: InputStream = connection.getInputStream()
        val isr = InputStreamReader(`in`)

        val listBuilder = DownloadableFormatListBuilder()
        val entries: MutableList<DownloadableFormatEntry>

        JsonReader(isr).use { reader ->
            entries = listBuilder.buildListFromJson(reader)
        }

        // Check for available updates
        val versionExtractor = DebateFormatFieldExtractor(mContext, R.string.xml2elemName_version)
        for (entry in entries)
            entry.checkForExistingFile(mFilesManager, versionExtractor)

        Collections.sort(entries)
        return Collections.unmodifiableList(entries)
    }

    @Throws(MalformedURLException::class)
    private fun verifyHostMatch(url: URL): Boolean {
        val listUrl = getListUrl()
        return url.host == listUrl.host
    }

    @Throws(MalformedURLException::class)
    private fun getListUrl(): URL {
        val prefs = PreferenceManager.getDefaultSharedPreferences(mContext)
        val userUrl =
                prefs.getString(DownloadFormatsFragment.PREFERENCE_DOWNLOAD_LIST_URL, null)
        return if (userUrl == null) {
            val defaultUrl = mContext.getString(R.string.formatDownloader_list_defaultUrl)
            Log.i(TAG, "using default URL: $defaultUrl")
            URL(defaultUrl)
        } else {
            Log.i(TAG, "using user-provided URL: $userUrl")
            URL(userUrl)
        }
    }
}
